File: dotnet/src/PqcStandards/MlKem/Sampling.cs
namespace PqcStandards.MlKem;

/// <summary>Sampling functions for ML-KEM (FIPS 203).</summary>
public static class Sampling
{
    /// <summary>SampleNTT: rejection sampling from XOF(rho || i || j).</summary>
    public static int[] SampleNtt(byte[] rho, byte i, byte j)
    {
        byte[] input = new byte[rho.Length + 2];
        Buffer.BlockCopy(rho, 0, input, 0, rho.Length);
        input[rho.Length] = i;
        input[rho.Length + 1] = j;

        // Generate enough XOF output
        byte[] buf = HashFuncs.Xof(input, 3 * 256 * 2);

        int[] a = new int[256];
        int ctr = 0;
        int pos = 0;
        while (ctr < 256)
        {
            int d1 = ((buf[pos] | (buf[pos + 1] << 8)) & 0xFFF);
            int d2 = (((buf[pos + 1] >> 4) | (buf[pos + 2] << 4)) & 0xFFF);
            pos += 3;

            if (d1 < Field.Q) a[ctr++] = d1;
            if (ctr < 256 && d2 < Field.Q) a[ctr++] = d2;
        }
        return a;
    }

    /// <summary>SamplePolyCBD: sample polynomial from centered binomial distribution.</summary>
    public static int[] SamplePolyCbd(byte[] sigma, byte nonce, int eta)
    {
        byte[] buf = HashFuncs.Prf(sigma, nonce, 64 * eta);
        int[] f = new int[256];

        for (int i = 0; i < 256; i++)
        {
            int a = 0, b = 0;
            for (int j = 0; j < eta; j++)
            {
                int bitIdx = 2 * i * eta + j;
                a += (buf[bitIdx / 8] >> (bitIdx % 8)) & 1;
            }
            for (int j = 0; j < eta; j++)
            {
                int bitIdx = 2 * i * eta + eta + j;
                b += (buf[bitIdx / 8] >> (bitIdx % 8)) & 1;
            }
            f[i] = Field.ModQ(a - b);
        }
        return f;
    }
}

File: go/internal/sampling/sampling.go
// Package sampling implements the sampling algorithms from FIPS 203 (ML-KEM).
package sampling

import (
	"github.com/liviuepure/PQC-Standards-Implementation/go/internal/field"
	"golang.org/x/crypto/sha3"
)

// SamplePolyCBD implements FIPS 203 Algorithm 8 (SamplePolyCBD_eta).
// It samples a polynomial from the centered binomial distribution.
// The input bytes must have length 64*eta.
func SamplePolyCBD(eta int, bytes []byte) [256]field.Element {
	var f [256]field.Element

	switch eta {
	case 2:
		// Each coefficient uses 2*eta = 4 bits total: 2 bits for x, 2 bits for y.
		// Process byte by byte; each byte yields 2 coefficients.
		for i := 0; i < 256; i += 2 {
			b := bytes[i/2]
			// First coefficient: bits 0,1 for x; bits 2,3 for y
			x := int16((b >> 0) & 1) + int16((b>>1)&1)
			y := int16((b >> 2) & 1) + int16((b>>3)&1)
			f[i] = field.FromI16(x - y)

			// Second coefficient: bits 4,5 for x; bits 6,7 for y
			x = int16((b >> 4) & 1) + int16((b>>5)&1)
			y = int16((b >> 6) & 1) + int16((b>>7)&1)
			f[i+1] = field.FromI16(x - y)
		}
	case 3:
		// Each coefficient uses 2*eta = 6 bits total: 3 bits for x, 3 bits for y.
		// We work at the bit level.
		bits := bytesToBits(bytes)
		for i := 0; i < 256; i++ {
			var x, y int16
			for j := 0; j < eta; j++ {
				x += int16(bits[6*i+j])
				y += int16(bits[6*i+eta+j])
			}
			f[i] = field.FromI16(x - y)
		}
	}

	return f
}

// bytesToBits converts a byte slice into a slice of individual bits (LSB first).
func bytesToBits(data []byte) []byte {
	bits := make([]byte, len(data)*8)
	for i, b := range data {
		for j := 0; j < 8; j++ {
			bits[i*8+j] = (b >> uint(j)) & 1
		}
	}
	return bits
}

// SampleNTT implements FIPS 203 Algorithm 7 (SampleNTT).
// It performs rejection sampling from SHAKE-128 output to produce
// a polynomial in NTT representation with coefficients in [0, q).
func SampleNTT(seed [34]byte) [256]field.Element {
	var a [256]field.Element

	h := sha3.NewShake128()
	h.Write(seed[:])

	buf := make([]byte, 3)
	j := 0
	for j < 256 {
		h.Read(buf)
		d1 := uint16(buf[0]) | (uint16(buf[1]&0x0F) << 8)
		d2 := uint16(buf[1]>>4) | (uint16(buf[2]) << 4)

		if d1 < field.Q {
			a[j] = field.Element(d1)
			j++
		}
		if d2 < field.Q && j < 256 {
			a[j] = field.Element(d2)
			j++
		}
	}

	return a
}

// PRF computes SHAKE-256(seed || nonce) and returns the first length bytes.
func PRF(seed [32]byte, nonce byte, length int) []byte {
	h := sha3.NewShake256()
	h.Write(seed[:])
	h.Write([]byte{nonce})

	out := make([]byte, length)
	h.Read(out)
	return out
}
